lib/Serialization/Serialization.cpp

```diff
@@ -3276,23 +3276,39 @@ class Serializer::DeclSerializer : public DeclVisitor<DeclSerializer> {
     return count;
   }
 
-public:
-  /// Determine if \p decl is safe to deserialize when it's public
-  /// or otherwise needed by the client in normal builds, this should usually
-  /// correspond to logic in type-checking ensuring these safe decls don't
-  /// refer to implementation details. We have to be careful not to mark
-  /// anything needed by a client as unsafe as the client will reject reading
-  /// it, but at the same time keep the safety checks precise to avoid
-  /// XRef errors and such.
-  static bool isDeserializationSafe(const Decl *decl) {
-    if (auto ext = dyn_cast<ExtensionDecl>(decl)) {
-      // Consider extensions as safe as their extended type.
+  class ExternallyAccessibleDeclVisitor
```

Comment:
I don't mind the name but I think we usually use "exported" for the idea of externally accessible. The term may already be overloaded however.

Reply:
Yeah, that's a good point. Maybe I should rename this and follow through on extracting this visitor into it's own file the hope that we can unify some or all of the other exportability checks in the compiler.

Reply:
There is a set of `isExported` functions that are also getting very close to this version as you're reducing the difference between was is serialization safe and just what's exported.

Reply:
I pulled the class out into a header and renamed it to `DeclExportabilityVisitor`.